Meson build system and CI workflow for automatic build testing

matsbn opened the pull request:
Meson is a promising build system for standalone BLOM application. The CI workflow automatically tests building on several systems and with different compilers. Both the Meson build system and CI workflow are non-intrusive and does not change BLOM operation when built as a NorESM component (the Meson branch compared to master gave bit-identical results in NorESM). A failed CI build test will not obstruct merging of commits, and therefore is merely a diagnostic of code portability (currently the macOS build tests fail because of a known Meson issue).

Commit:
Added Meson build system

Commit:
Added Meson build system

Commit:
Moved building of 'dimensions.F' into own file

Commit:
Moved building of 'dimensions.F' into own file

Commit:
Updated README with build instructions for Meson

Commit:
Updated README with build instructions for Meson

matsbn committed:
Added support for macOS, functionality for having the build directory elsewhere than in the source root and modified some variable names and comments related to the construction of dimensions.F.

matsbn committed:
Added support for macOS, functionality for having the build directory elsewhere than in the source root and modified some variable names and comments related to the construction of dimensions.F.

Commit:
Initial Github action setup

Commit:
Updated to test with default OS

Commit:
Fixed name for 'pull_request'

Commit:
Fixed syntax error

Commit:
Fixed name error for 'macos' and moved to 'apt-get'

Commit:
Updated to manual install of Meson

Commit:
Install meson globally?

Commit:
Update to CI

Commit:
Updated compiler handling on macOS

Commit:
Simplified runners

Commit:
Updated version of Meson for Ubuntu CI

Commit:
Experiment to test different GCC version on macOS

Commit:
Reverted macOS test

Commit:
Fixed 'dimensions.F' generation on Linux

Commit:
Added CI job with Intel compilers

Commit:
Changed Intel docker image, hopefully light enough

Commit:
Simplified syntax, added upload of meson log

Commit:
Change to 'matrix.os' for build artifacts

Commit:
Changed to OneAPI Ubuntu container

Commit:
Force container to update dependencies

Commit:
Experiment to find correct Fortran package

Commit:
Change Intel compiler package

Commit:
Changed to native Linux for Intel compiler

Commit:
Run Intel job concurrently with build step

Commit:
Added explicit install of Intel Fortran compiler

Commit:
Source Intel before compile step

Commit:
Find correct netCDF for Intel compilers

Commit:
Add custom compile of netCDF for Intel compilers

Commit:
Change to 'add_project_arguments' instead of global

Commit:
Fixed repository path for netCDF

Commit:
Test compile netCDF

Commit:
Removed 'working-directory'

Commit:
Removed change directory

Commit:
C version of netCDF is required to build Fortran version

Commit:
Downgrade netCDF Fortran version

Commit:
Removed 'make install', requires 'sudo'

Commit:
Don't 'make check' netCDF

Commit:
Cleanup

Commit:
Fixed script to create 'dimensions.F' on Linux

It seems that 'read -d " "' behaves differently on macOS and Linux
(tested on Ubuntu 20.04 and Fedora 33) and on Linux will output empty
lines which don't appear on macOS. This change adds 'if' clauses to
ignore empty reads. The change has been tested on Linux and macOS and
works on both.

Commit:
Fixed script to create 'dimensions.F' on Linux

It seems that 'read -d " "' behaves differently on macOS and Linux
(tested on Ubuntu 20.04 and Fedora 33) and on Linux will output empty
lines which don't appear on macOS. This change adds 'if' clauses to
ignore empty reads. The change has been tested on Linux and macOS and
works on both.

Commit:
Rebased to latest master

Commit:
Merge with master

Commit:
Merge branch 'meson' into github_action

Commit:
Added testing of MPI build

Commit:
Fixed matrix arguments for MPI

Commit:
Reference matrix.mpi instead of env.mpi

Commit:
Add missing MPI packages

Commit:
Fix gfortran argument mismatch

Commit:
Change branch for CI in preparation for upstream

matsbn committed:
Modified script creating dimensions.F.

Element-wise 'read' is replaced with array read.

nordmoen committed:
Merge pull request #64 from nordmoen/github_action

Add CI integration with Github Actions

Commit:
Added support for compiling hamocc with Meson

Commit:
Merge branch 'master' of https://github.com/JorgSchwinger/BLOM into meson_hamocc

Commit:
Merge branch 'master' of https://github.com/JorgSchwinger/BLOM into meson_hamocc

Commit:
Remove compiler flag to allow long lines

Removed compiler flags after changes from PR #67

nordmoen committed:
Merge pull request #66 from nordmoen/meson_hamocc

Added support for compiling hamocc with Meson

matsbn committed:
Added gcc and intel compiler debug flags for Meson build.

matsbn committed:
Merge pull request #68 from matsbn/meson

Added gcc and intel compiler debug flags for Meson build.

nordmoen commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,117 @@
+on:
+  push:
+    branches:
+      - meson
```
This should probably be changed to `master` so that CI runs on pushes to `master` and not the `meson` branch

nordmoen commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,117 @@
+      - meson
+  pull_request:
+    branches:
+      - meson
```
This should also be changed to `master` (see comment above)

matsbn replied:
Thank you for pointing this out. I have committed a modified .github/workflows/ci.yml so the tests will be done on master branch.

matsbn committed:
Make sure CI runs on pushes to master branch.

matsbn committed:
Merge remote-tracking branch 'upstream/meson' into meson

nordmoen commented:
It is also possible to simplify the CI configuration a bit ([ref](https://github.com/nordmoen/BLOM/blob/meson/.github/workflows/ci.yml)) which would avoid the failing tests on macOS, but that can be a new feature after this pull request has been merged.

TomasTorsvik committed:
Update README.md

Maybe indicate that meson is an option for stand-alone BLOM, not for the NorESM system.

TomasTorsvik reviewed (approved):
I made a small change in the README.md file. Should have included this as a suggestion, but I messed it up a bit. Still, it's should be just a minor thing. Looks fine otherwise.

AleksiNummelin reviewed (approved):
Looks good! Latest Ubuntu build seems to be failing now, but I think that is acceptable (and might not have anything to do with the build system).
